### DISCUSSION TITLE
Is it intentional that you can't attach scripts that aren't toplevel

### DISCUSSION THREAD

**Discussion by ScartCable** [HaxeGodot/godot#7] [General]
If I place a .hx file in a package 'name' the compiler will give out a error saying that `package name; should be package;`. The error comes from the godot.Godot file so i'm just wondering if this intentional? It seems intentional due to error originating from godot.Godot macro processing.

Also another unrelated error. This question may be better suited for the haxe forum but I though I'd ask here first because it is your project. In my haxe godot project macros simply do not work. Build macros turnoff autocompletion, expression macro args and returns are always null and if I have a macro that just embeds a string the code will refuse to compile due to ` You cannot access the cs package while in a macro (for cs.NativeArray)`.

**Comment by ibilon** [comment]
You can have scripts that aren't top level, is your file in the `scripts/name` folder?
This error means "you said this is in the name package but I found it in the top level package" so either the package line or the file path is wrong.
Scripts run a build macro from `godot.Godot` to make the metas like `@:export` work, so this might be why the error seems to come from there.

For your macro error it means you are importing types in the cs package from a file in macro mode, all godot imports end up importing the cs package.
So you need to either wrap them in a `#if !macro (imports) #end` block,
or a another way (I'd say better way) would be to have your macro code in its own file (as much as possible) this avoid some errors like this.

Though maybe it's not coming from your code, the file `scripts/import.hx` has some imports, maybe they need to be wrapped in a `#if !macro` block.

**Reply by ScartCable** [reply]
You were right about import.hx messing with the macro files, totally forgot about import. Thanks!

So about the package thing. With the file lying like this: script/name/Name.hx and Name.hx looking like this:
```
package name;
class Name {
  // stuff
}
```

The compiler will emit these errors as soon as the file is attached to a node (but not before)

`package name;` in scripts/name/Name.hx should be `package;`
Module "name/Name" does not have a valid name. Module name should start with an uppercase letter: "name/Name"
Module name/Name does not define type name/Name

if I remove the `package name;` only the two other errors remain. The errors seems to originate from `Godot.buildProject();` because the compiler says that the error comes from (unknown) and running a global diagnostics check in vscode places the errors under `--macro godot.Godot.buildProject`.

**Reply by ibilon** [reply]
What version of the externs are you using? There were errors like that on a earlier version, but then fixed.

**Reply by ScartCable** [reply]
I'm using 3.3.0. I just did `haxelib install godot`.
